src/modules/types.rs

```diff
@@ -14,14 +14,36 @@ pub enum Type {
     Generic
 }
 
+impl Type {
+    pub fn is_subset_of(&self, other: &Type) -> bool {
+        match (self, other) {
+            (Type::Array(left), Type::Array(right)) => {
+                **left != Type::Generic && **right == Type::Generic
```

Comment:
what if its the opposite? like `left = [] ; right = [T]`

Reply:
Then the left type is not a subset of the right type. I've updated variable names to make it clearer.